src/utils/config-context.ts

```diff
@@ -8,33 +14,38 @@ import {
   SWRConfiguration,
   FullConfiguration,
   ProviderConfiguration,
-  Cache
+  Cache,
+  CacheProvider
 } from '../types'
 
 export const SWRConfigContext = createContext<Partial<FullConfiguration>>({})
 
-const SWRConfig: FC<{
-  value?: SWRConfiguration &
-    Partial<ProviderConfiguration> & {
-      provider?: (cache: Readonly<Cache>) => Cache
-    }
-}> = ({ children, value }) => {
+type SWRConfigValue = SWRConfiguration &
+  Partial<ProviderConfiguration> & {
+    provider?: CacheProvider<any>
```

Comment:
This does not look correct to me.
```typescript
export type Cache<T = DefaultProvider, Data = any> = T extends ICache<Data>
  ? T
  : never

export type CacheProvider<Data = any> = (cache: Cache<Data>) => Cache<Data>

CacheProvider<any> = (cache: Cache<any>) => Cache<any> = (cache: any) => any
```
We need provider to return a Cache here

Reply:
I didn't get what do you mean here. Can you rephrase it more specificly?

Reply:
## Previous
```typescript
type Provider: (cache: Readonly<Cache>) => Cache
const cache = new Map()
<SWRConfig value={{ provider: () => cache }}>{children}</SWRConfig>
// This will throw type error since { a: foo } does not satisfy the requirement
<SWRConfig value={{ provider: () => {a: foo } }}>{children}</SWRConfig>
```

## Current
```typescript
type Provider = CacheProvider<any> = (_: any) => any
const cache = new Map()
<SWRConfig value={{ provider: () => cache }}>{children}</SWRConfig>
// This will not throw type error any more
<SWRConfig value={{ provider: () => {a: foo } }}>{children}</SWRConfig>
```